grpc/src/main/java/com/linecorp/armeria/server/grpc/DefaultUnframedGrpcErrorHandler.java

```diff
@@ -0,0 +1,276 @@
+                                                                           grpcCode.value())
+                                                                   .build();
+            final HttpData content;
+            try (TemporaryThreadLocals ttl = TemporaryThreadLocals.acquire()) {
+                final StringBuilder msg = ttl.stringBuilder();
+                msg.append("grpc-code: ").append(grpcCode.name());
+                if (grpcMessage != null) {
+                    msg.append(", ").append(grpcMessage);
+                }
+                if (cause != null && ctx.config().verboseResponses()) {
+                    msg.append("\nstack-trace:\n").append(Exceptions.traceText(cause));
+                }
+                content = HttpData.ofUtf8(msg);
+            }
+            return HttpResponse.of(responseHeaders, content);
+        };
+    }
+
+    /**
+     * Returns a rich error JSON response based on Google APIs.
+     * Please refer <a href="https://cloud.google.com/apis/design/errors#error_model">Google error model</a>
+     * @param statusMappingFunction The function which maps the {@link Throwable} or gRPC {@link Status} code
+     *                              to an {@link HttpStatus} code.
+     */
+    static UnframedGrpcErrorHandler ofRichJson(UnframedGrpcStatusMappingFunction statusMappingFunction) {
+        final UnframedGrpcStatusMappingFunction mappingFunction =
+                requireNonNull(statusMappingFunction, "statusMappingFunction")
+                        .orElse(UnframedGrpcStatusMappingFunction.of());
+        return (ctx, status, response) -> {
+            final Code grpcCode = status.getCode();
+            @Nullable
+            final String grpcMessage = status.getDescription();
+            @Nullable
+            final Throwable cause = getThrowableFromContext(ctx);
+            final HttpStatus httpStatus = mappingFunction.apply(ctx, status, cause);
+            final HttpHeaders trailers = !response.trailers().isEmpty() ?
+                                         response.trailers() : response.headers();
+            @Nullable
+            final String grpcStatusDetailsBin = trailers.get(GrpcHeaderNames.GRPC_STATUS_DETAILS_BIN);
+            final ResponseHeaders responseHeaders = ResponseHeaders.builder(httpStatus)
+                                                                   .contentType(MediaType.JSON_UTF_8)
+                                                                   .addInt(GrpcHeaderNames.GRPC_STATUS,
+                                                                           grpcCode.value())
+                                                                   .build();
+            final ImmutableMap.Builder<String, Object> messageBuilder = ImmutableMap.builder();
+            messageBuilder.put("code", httpStatus.code());
+            messageBuilder.put("status", grpcCode.name());
```

Comment:
Should we include the `status`? By the definition from the doc, status is a `Message` that has `code`, `message` and `details`.

Reply:
> it seems like the official [google document](https://developers.google.com/api-client-library) includes status.

I couldn't find the status in the link you gave. 😅 Could you give me a detailed link to it? 🤔

Reply:
Sorry I attached a wrong link. This [error document](https://cloud.google.com/apis/design/errors#error_model) gives us an example
```
{
  "error": {
    "code": 400,
    "message": "API key not valid. Please pass a valid API key.",
    "status": "INVALID_ARGUMENT",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.ErrorInfo",
        "reason": "API_KEY_INVALID",
        "domain": "googleapis.com",
        "metadata": {
          "service": "translate.googleapis.com"
        }
      }
    ]
  }
}
```
It says 
`While proto3 messages have native JSON encoding, Google's API Platform uses a different error schema for Google's JSON HTTP APIs for backward compatibility reasons.`

Reply:
> While proto3 messages have native JSON encoding, Google's API Platform uses a different error schema for Google's JSON HTTP APIs for backward compatibility reasons.

I interpret this to mean that
1) `proto3` messages have natively defined [json-encoding](https://developers.google.com/protocol-buffers/docs/proto3#json). I think this makes sense 
2) Due to backwards compatibility, google's API Platform uses a different schema for backwards compatibility reasons.

While I agree it can be useful to have a `Google API Platform` variant of `UnframedGrpcErrorHandler`, I'm a little concerned that this variant isn't backwards compatible with proto3's json encoding since the `code` value would be encoded differently.

For instance, envoy seems to use the enum value for the `code` value: https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/grpc_json_transcoder/v3/transcoder.proto

<img width="997" alt="Screen Shot 2022-06-16 at 4 37 31 PM" src="https://user-images.githubusercontent.com/8510579/174017969-bc2fd680-62ef-41b6-a4d4-d58ab23aa920.png">

I'm wondering 1) do we necessarily want to support the `Google API Platform` version 2) if we do decide to support this, do we want to define a separate factory method for this? e.g. `UnframedGrpcErrorHandler#ofRichJson` vs. `UnframedGrpcErrorHandler#ofRichJson(googleApiPlatformVersion=true)` <- this is a bad name, but something like this 😅

Reply:
I am leaning toward using a more standard value of `code` defined in the specification.
https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto#L36-L37
As @jrhee17 mentioned, we can provide a compatible version with Google API Client later on request.

Reply:
> > While proto3 messages have native JSON encoding, Google's API Platform uses a different error schema for Google's JSON HTTP APIs for backward compatibility reasons.
> 
> I interpret this to mean that
> 
> 1. `proto3` messages have natively defined [json-encoding](https://developers.google.com/protocol-buffers/docs/proto3#json). It is fine to use this to encode `Status`
> 2. Due to backwards compatibility, google's API Platform uses a different schema for backwards compatibility reasons.
> 
> While I agree it can be useful to have a `Google API Platform` variant of `UnframedGrpcErrorHandler`, I'm a little concerned that this variant isn't backwards compatible with proto3's json encoding since the `code` value would be encoded differently.
> 
> For instance, envoy seems to use the enum value for the `code` value: https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/grpc_json_transcoder/v3/transcoder.proto
> 
> <img alt="Screen Shot 2022-06-16 at 4 37 31 PM" width="997" src="https://user-images.githubusercontent.com/8510579/174017969-bc2fd680-62ef-41b6-a4d4-d58ab23aa920.png">
> 
> I'm wondering
> 
> 1. Do we necessarily want to support the `Google API Platform` specific version? I'm curious about your use-case 🙏  I'm all for making life easier for `armeria` users, but I first want to make sure this is the exact behavior you are targeting.
> 
> If we do decide to support this, I guess we may want to define a separate factory method for this e.g. `UnframedGrpcErrorHandler#ofRichJson` vs. `UnframedGrpcErrorHandler#ofRichJson(googleApiPlatformVersion=true)` <- this doesn't have to be done in this PR, but I imagine we might end up needing to support both formats.

1. I think we do not have to support Google API Platform specific version. My intension is to make Armeria have a nicer error handling which some other frameworks like envoy and grpc-gateway already have which can easily translate grpc's native error library into http error message.
2. I am not quite sure what you mean. But as long as we keep our error handling backwards compatible we do not have to worry about google api version. Google seems to change their error handling a lot so we don't have to be exactly the same with google?

Reply:
I prefer to use http status code. After all, the http code might be more intuitive for the clients like app or webapp since it is more likely that they are more familiar with http standards?

Reply:
I've also checked grpc-gateway, and it also seems to return the grpc code as opposed to the http status code. (the same behavior as just `proto3`'s `json-transcoding` and `envoy-proxy`)

envoy
```
...
< HTTP/1.1 400 Bad Request
< content-type: application/json
< grpc-encoding: identity
< grpc-accept-encoding: gzip
< content-length: 61
< server: envoy
< date: Fri, 17 Jun 2022 07:20:08 GMT
< x-envoy-upstream-service-time: 16
< 
{
 "code": 3,
 "message": "sample message",
 "details": []
}
```

grpc-gateway
```
...
< HTTP/2 429 
< content-type: application/json
< content-length: 52
< date: Fri, 17 Jun 2022 09:23:41 GMT
< 
* Connection #0 to host 0.0.0.0 left intact
{"code":8, "message":"custom message", "details":[]}%     
```

> the http code might be more intuitive for the clients like app or webapp since it is more likely that they are more familiar with http standards?

I agree with this point, but I also think this is duplicate information since the "http code" will also be returned as the http status code anyways.

e.g.
```
...
< HTTP/2 429 
...
{"code":429, "message":"custom message", "details":[]}%     
```

If we decide to use `grpc.code` as the code value, we can also remove the `status` field since this is also duplicate information.
I think I prefer to have a default behavior which matches the other proxy/gateway behaviors.

I'm also curious what other maintainers think

Reply:
That's reasonable. I can change the code to `grpc.code` instead of http code. @ikhoon @minwoox @trustin What do you think?

Reply:
> If we decide to use grpc.code as the code value, we can also remove the status field since this is also duplicate information.

I agree with @jrhee17's idea. If our response format is compatible with other open-source such as grpc-gateway or Envoy, users can smoothly migrate to Armeria or vice versa.

Reply:
It makes sense that we make Armeria's error model compatible to popular open-source platforms. I just updated the response code to grpc.code and remove the status field.